### PR TITLE
Feat/rust ast check

### PR DESCRIPTION
**Pull request by yzhang1991**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: libflux/src/ast/check/mod.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+use crate::ast::SourceLocation;
+use std::fmt;
+
+#[cfg(test)]
+mod tests;
+
+// check will inspect a single node and annotate it with any AST errors.
+// Unlike the ast.Check() in our Go parser, the check function in Rust
+// returns a list of errors generated during the check process PLUS
+// any errors existed before ast.check() is performed.
```
Comment (nathanielc) [inline libflux/src/ast/check/mod.rs]: nit, Can you update this comment to describe what it does without comparing it to the Go check function? We want this comment to serve as the documentation for the function and its likely that who ever is reading it isn't familiar with the Go behavior.

File: libflux/src/ast/check/tests.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            },
+            end: Position {
+                line: 2,
+                column: 14
+            },
+            source: Some(String::from("{c: 2, a}"))
+        },
+        message: String::from("cannot mix implicit and explicit properties")
+    }];
+    assert!(want.iter().zip(got).all(|(want, got)| want == &got));
```
Comment (nathanielc) [inline libflux/src/ast/check/tests.rs]: if you change this to `assert_eq!(want, got)` does it work? If so I would prefer that method as Rust assert_eq method has good diffing logic for what is different between the values.

File: libflux/src/ast/check/mod.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                    location: n.base().location.clone(),
+                    message: err.clone()
+                });
+            }
+            match n {
+                walk::Node::BadStmt(n) => {
+                    errors.push(Error {
+                        location: n.base.location.clone(),
+                        message: format!("invalid statement {}: {}",
+                                         n.base.location, n.text)
```
Comment (nathanielc) [inline libflux/src/ast/check/mod.rs]: Since we now store the location on the error type itself I don't think we want it in the error message. Thoughts?

Comment (yzhang1991) [reply libflux/src/ast/check/mod.rs]: Makes sense.

File: libflux/src/ast/check/mod.rs (hunk 1 above)
Comment (yzhang1991) [reply libflux/src/ast/check/mod.rs]: Done

File: libflux/src/ast/check/tests.rs (hunk 2 above)
Comment (yzhang1991) [reply libflux/src/ast/check/tests.rs]: Done

**Pull request merged**